### DISCUSSION TITLE
useContractRead return error when gettting erc20 token balance

### DISCUSSION THREAD

**Discussion by LuiJunYe** [wevm/wagmi#2834] [Q&A] [answered]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package Version
wagmi  : 1.3.9
viem : 1.4.2

### Currenct Behaviour
when calling useContractRead hooks to get balance for an erc20 token using 0xstring as address, return value is undefined. an error shows that address provided in 0xstring is invalid

### Expected Behaviour
return erc20 token balance

### Steps To Reproduce
```
import { erc20ABI, useContractRead } from 'wagmi'

 const { data } = useContractRead({
   chainId: 1,
   abi: erc20ABI,
   address: 0x0xdAC17F958D2ee523a2206206994597C13D831ec7,
   functionName: 'balanceOf',
   args: [0xaddress]
 })

```

### Anything else
The following is the error message block that throw the error

```
Docs: https://viem.sh/docs/contract/readContract.html
Version: viem@1.4.2
    at getContractError (getContractError.ts:55:1)
    at readContract (readContract.ts:100:1)
    at async Object.queryFn (index.js:1243:1)
```

**Comment by SaadNiaziDev** [comment]
Your contract address is invalid. it should be `0xdAC17F958D2ee523a2206206994597C13D831ec7`

**Reply by LuiJunYe** [reply]
based on the library file on viem, it is require to change to 0xAddress. i have try to remove the 0x but got error.

```
export type ContractFunctionConfig<TAbi extends Abi | readonly unknown[] = Abi, TFunctionName extends string = string, TAbiStateMutability extends AbiStateMutability = AbiStateMutability> = {
    /** Contract ABI */
    abi: Narrow<TAbi>;
    /** Contract address */
    address: Address;
    /** Function to invoke on the contract */
    functionName: InferFunctionName<TAbi, TFunctionName, TAbiStateMutability>;
} & GetFunctionArgs<TAbi, TFunctionName>;
```

**Comment by SaadNiaziDev** [comment]
try by passing address as string.

`const { data } = useContractRead({
   chainId: 1,
   abi: erc20ABI,
   address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
   functionName: 'balanceOf',
   args: [0xaddress]
 })`

**Reply by LuiJunYe** [reply]
Passing address as string works.
 
i have check also that all useContract function is the same as well. if need pass address as string then how to dynamically read/write contract. can provide suggestions on this ?

i have try  .toString as well but it seams not working as well.

**Reply by SaadNiaziDev** [reply]
Dynamic value would be string by default so why you want to use .toString ? Just pass as it is. If issue persist then use template literals `` `${your address}` ``

**Reply by LuiJunYe** [reply]
i have try using template literals but will get validation error

```
Type 'string' is not assignable to type '`0x${string}`'
```

**Reply by petermazzocco** [reply, marked as answer]
Import the type Address and pass it as Address like:

`import type { Address } from "viem";`

And then do:

```
const { data } = useContractRead({ chainId: 1, abi: erc20ABI, address: address as Address, functionName: 'balanceOf', args: [otherAddr as address] })
```

Please mark as answered when you've set it up :)

**Reply by LuiJunYe** [reply]
Thanks for the help, it works

**Reply by LuiJunYe** [reply]
Thanks all for helping
